### PR TITLE
Feature/update pr829

### PR DESCRIPTION
**Pull request by adev4a**
# PR Details
This pr updates the identify lane change functionality in the route_following_plugin and also implements the changes from pr#829 following these updates.
## Description

The identifyLaneChange function boolean return values are reversed. A lane change logic is added to route_following, a maximum distance required for lane change is calculated and for the distance before that, a lane following maneuver is requested.
## How Has This Been Tested?

The code was unit tested and integration tested.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: route_following_plugin/include/route_following_plugin.h [hunk 1]
```diff
@@ -77,7 +91,11 @@ namespace route_following_plugin
          * \param current_time Start time of the current maneuver
          * \return A lane keeping maneuver message which is ready to be published
          */
-        cav_msgs::Maneuver composeStopandWaitManeuverMessage(double current_dist, double end_dist, double current_speed, int start_lane_id, int target_lane_id, ros::Time current_time, double end_time);
+        cav_msgs::Maneuver composeStopandWaitManeuverMessage(double current_dist, double end_dist, double current_speed, int start_lane_id, int target_lane_id, ros::Time& current_time, double end_time);
+
+        cav_msgs::Maneuver composeLaneChangeManeuverMessage(double current_dist, double end_dist, double current_speed, double target_speed, int starting_lane_id,int ending_lane_id, ros::Time& current_time);
```
Comment (saina-ramyar) [inline route_following_plugin/include/route_following_plugin.h:96]: add description

File: route_following_plugin/include/route_following_plugin.h [hunk 2]
```diff
@@ -77,7 +91,11 @@ namespace route_following_plugin
...
+
+        void updateCurrentStatus(cav_msgs::Maneuver maneuver, double& speed, double& end_dist, int& lane_id);
```
Comment (saina-ramyar) [inline route_following_plugin/include/route_following_plugin.h:98]: add description

File: route_following_plugin/src/route_following_plugin.cpp [hunk 3]
```diff
@@ -122,10 +134,41 @@ namespace route_following_plugin
             if(end_dist < current_progress){
                 break;
             }
-            resp.new_plan.maneuvers.push_back(
+            
+            auto following_lanelets = wm_->getRoute()->followingRelations(shortest_path[last_lanelet_index]);
+
+            if(last_lanelet_index!= (shortest_path.size()-1) && identifyLaneChange(following_lanelets, shortest_path[last_lanelet_index + 1].id()))
```
Comment (saina-ramyar) [inline route_following_plugin/src/route_following_plugin.cpp:140]: add a comment for purpose of this if condition

File: route_following_plugin/src/route_following_plugin.cpp [hunk 4]
```diff
@@ -203,9 +236,38 @@ namespace route_following_plugin
...
+        maneuver_msg.lane_change_maneuver.parameters.presence_vector = cav_msgs::ManeuverParameters::HAS_TACTICAL_PLUGIN;
+        maneuver_msg.lane_change_maneuver.parameters.planning_tactical_plugin = "UnobstructedLaneChangePlugin";
+        maneuver_msg.lane_change_maneuver.parameters.planning_strategic_plugin = "RouteFollowingPlugin";
+        maneuver_msg.lane_change_maneuver.start_dist = current_dist;
+        maneuver_msg.lane_change_maneuver.start_speed = current_speed;
+        maneuver_msg.lane_change_maneuver.start_time = current_time;
+        maneuver_msg.lane_change_maneuver.end_dist = end_dist;
+        maneuver_msg.lane_change_maneuver.end_speed = target_speed;
+        // because it is a rough plan, assume vehicle can always reach to the target speed in a lanelet
+        double cur_plus_target = current_speed + target_speed;
```
Comment (saina-ramyar) [inline route_following_plugin/src/route_following_plugin.cpp:257]: not sure what is the purpose of this variable

Comment (adev4a) [reply route_following_plugin/src/route_following_plugin.cpp:257]: adopted from compose_lane_follow

**Pull request merged**